internal/pkg/agent/storage/store/migrations.go

```diff
@@ -0,0 +1,201 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package store
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage/store/internal/migrations"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+var ErrInvalidYAML = errors.New("could not parse YAML")
+
+func migrateActionStoreToStateStore(
+	log *logger.Logger,
+	actionStorePath string,
+	stateDiskStore storage.Storage) (err error) {
+
+	log = log.Named("state_migration")
+	actionDiskStore, err := storage.NewDiskStore(actionStorePath)
+	if err != nil {
+		return fmt.Errorf(
+			"could not create disk store when migratins action store: %w", err)
+	}
+
+	stateStoreExists, err := stateDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if state store exists: %v", err)
+		return err
+	}
+
+	// do not migrate if the state store already exists
+	if stateStoreExists {
+		log.Debugf("state store already exists")
+		return nil
+	}
+
+	actionStoreExists, err := actionDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if action store %s exists: %v", actionStorePath, err)
+		return err
+	}
+
+	// delete the actions store file upon successful migration
+	defer func() {
+		if err == nil && actionStoreExists {
+			err = actionDiskStore.Delete()
+			if err != nil {
+				log.Errorf("failed to delete action store %s exists: %v", actionStorePath, err)
+			}
+		}
+	}()
+
+	// nothing to migrate if the action store doesn't exist
+	if !actionStoreExists {
+		log.Debugf("action store %s doesn't exists, nothing to migrate", actionStorePath)
+		return nil
+	}
+
+	action, err := migrations.LoadActionStore(actionDiskStore)
+	if err != nil {
+		log.Errorf("failed to create action store %s: %v", actionStorePath, err)
+		return err
+	}
```

Comment:
>[...] the action store gets deleted when this function returns, meaning the next trip through this function will succeed.
>
>If this is the correct behavior, can we just treat this error as non-fatal and move on?

It only gets deleted on success, so no, it will not succeed on the next run.

Reply:
Yes I read the behavior backwards for some reason, but it doesn't work out as nicely for this comment.

This adds a new way for the agent to get stuck at init time, which as mentioned worries me. I'd prefer if we can design this failure mode out of the system.

Reply:
https://github.com/elastic/elastic-agent/pull/4441#issuecomment-2038402088